### ISSUE TITLE
Need a way to load specific symbols

### ISSUE THREAD

**Issue by rikkimax** [BindBC/bindbc-opengl#3]
In SPEW, context creation requires the ability to lookup specific symbols (supports compat since it gets messy cross platform).

Currently with bindbc-opengl the ability to load specific functions are hidden behind private.
In essence I need access to bindGLSymbol in the form of ``void* function(string)``.

Would this be possible or will I need to write up a custom loader as a backup?

**Comment by mdparker** [comment]
That's probably a good idea. It would also be useful for anyone who wants to load a specific extension themselves if I haven't implemented it yet.

**Comment by mdparker** [comment]
Actually, it needed more than just the context check to handle it properly. However, the little test app I have keeps crashing when I insert a call to the current implementation in my little test app and I've been unable to debug it in the limited time I have today.  I can get to it again this weekend, but if you can try it and see if it works for your use case, that will help.

**Comment by rikkimax** [comment]
I need access to symbols like ``wglCreateContext``, which means that there very well may not be a context. So as long as ``loadOpenGL`` has been called once and ``bindGLSymbol`` will support both dlsym and ``getProcAddress`` I'm good. Patch incoming.

**Comment by mdparker** [comment]
Then you need to go through the normal loader and not through `bindGLSymbol'.  The context creation functions should be loaded directly through `LoadLibrary` and `GetProcAddress`, which is what the generic l loader uses.

**Comment by mdparker** [comment]
In other words, you don't need a special function in the gl binding. Go through bindbc-loader to get a handle to OpenGL32.dll and load the functions you need.

**Comment by rikkimax** [comment]
Not all context creation functions are available via ``dlsym``, these are primarily for the legacy functions cross-platform. For the newer contexts along with their housekeeping functions you do need to go through the specific ``getProcAddress`` function as an addition to the ``dlsym`` ones because the newer functions are extensions.

Problem is, for context creation such as in SPEW, the choice on which to use is two fold. What the user wants and what the platform supports. As such, both need to be tried if requested (for the newer context, legacy is pretty restricted to just < 3.2).

For X11 I also need access to ``glGetString`` because of some weird bugs that can crop up for e.g. XWayland. This really makes things really difficult from a library author's perspective.

So yes, I need to to be able to go through the logic of ``bindGLSymbol``. With the additional ``dlsym`` lookup. This is not optional. The question is, where is the code located? And how opt-in is it?

For reference here is the code snippet (from example file which spawned this issue) which works with my current PR:

```D
void setupOpenGL(managed!IWindowCreator creator) {
    import bindbc.opengl.gl;
    import bindbc.opengl.context;

    static OpenGL_Context_Callbacks callbacks;

    if (callbacks is OpenGL_Context_Callbacks.init) {
        callbacks.onLoad = (string platformLoadFunction) { loadOpenGL(); };
        callbacks.onReload = () { loadOpenGL(); };

        callbacks.loadSymbol = (string name) {
            assert(name.length < 256);

            char[256] temp;
            temp[0 .. name.length] = name;
            temp[name.length] = '\0';

            void* ptr;
            loadOpenGLSymbol(&ptr, temp.ptr);
            return ptr;    
        };

        loadOpenGL();
    }

    creator.assignOpenGLContext(OpenGLVersion(2, 1), &callbacks);
}
```

**Comment by mdparker** [comment]
Okay. Been a long time since I mucked around with the system APIs for context creation. I've removed your `loadOpenGLSymbol` function and added a new module, `bindbc.opengl.util`. This is not in `package.d`, so you'll need to import it directly. In it you'll find two functions: `loadBaseGLSymbol` and `loadExtendedGLSymbol`. I think it's best to keep them separate and leave the logic of whether to try one or both to the caller. Let me know if this suits your purposes.

```d
import bindbc.opengl.util;
pglGetString myGetString;
assert(loadBaseGLSymbol(cast(void**)&myGetString, "glGetString"));

pglShaderBinary myShaderBinary;
assert(loadExtendedGLSymbol(cast(void**)&myShaderBinary, "glShaderBinary"));
```
